### PR TITLE
Demo

### PR DESCRIPTION
**Pull request by LeviHHH**

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -98,34 +110,19 @@ type TokenWeight @entity {
...
-    "transitioning"
-    "expired"
+    active
+    transitioning
+    expired
 }
 
 type System @entity {
     "pricing and constant"
-    """
```
Comment (LeviHHH) [inline schema.graphql:108]: seem like it doesn't support multi-line comments

File: subgraph.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        - name: BucketFactory
+          file: ./abis/BucketFactory.sol/BucketFactory.json
+      eventHandlers:
+        - event: Create(address,address,uint256,uint256,uint256,string)
+          handler: handleCreate
+  - name: Bucket
+    kind: ethereum/contract
+    network: mainnet
+    source:
+      abi: Bucket
```
Comment (wfubuild) [inline subgraph.yaml:52]: should this be BucketFactory?

File: schema.graphql [hunk 3]
```diff
@@ -1,10 +1,9 @@
+"https://thegraph.com/docs/en/developer/create-subgraph-hosted/"
 type Account @entity {
     "user information"
     id: ID!
-    address: String!
```
Comment (JackyWYX) [inline schema.graphql:4]: Why address is not needed?

File: schema.graphql [hunk 4]
```diff
@@ -37,23 +34,38 @@ type Bucket @entity {
...
 
     createTime: Int!
     rounds: [BucketRound!] @derivedFrom(field: "bucket")
 }
 
+type SearchTicket @entity {
+    id: ID!
+    bucketRound: BucketRound!
+    user: Account!
+    accumulateTicktes: Int!
```
Comment (JackyWYX) [inline schema.graphql:48]: Spelling

File: schema.graphql [hunk 5]
```diff
@@ -37,23 +34,38 @@ type Bucket @entity {
...
     blockchain: String!
-    launchDate: Date!
+    launchDate: Int!
     referral: String
 
     createTime: Int!
     rounds: [BucketRound!] @derivedFrom(field: "bucket")
 }
 
+type SearchTicket @entity {
```
Comment (JackyWYX) [inline schema.graphql:44]: Thanks for adding thiese fields.

File: schema.graphql [hunk 6]
```diff
@@ -37,23 +34,38 @@ type Bucket @entity {
...
     state: RoundState!
     bucket: Bucket!
     round: Int!
     tickets: Int!
-
-    raffleTime: Date
-    txHash: String
+    raffleTime: Int
+    searchLength: Int!
+    searchTickets: [SearchTicket!] @derivedFrom(field: "bucketRound")
```
Comment (JackyWYX) [inline schema.graphql:67]: What does it mean by searchTickets?

File: schema.graphql (hunk 1 above)
Comment (JackyWYX) [reply schema.graphql:108]: OK. Is there any other way to add the comments?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+## 1. setup contract enviroment
+1. clone contract project and switch to 'fullTest' branch.
+```
+git clone https://github.com/nft-squared/whitelist-contracts
+cd whitelist-contracts
+git checkout origin/fullTest -b fullTest
+```
+2. install: `npm insall` or `yarn`
+3. start a local evm node: `yarn enode`
+4. deploy contracts and do some test: `yarn test --network develop test/FullLogic.test.ts`
```
Comment (JackyWYX) [inline README.md:13]: yarn compile

File: src/Bucket.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        if(!userRoundTicket) {
+            userRoundTicket = new UserRoundTicket(userRoundId)
+            userRoundTicket.bucketRound = round.id
+            userRoundTicket.amount = 0
+            round.userTickets.push(userRoundTicket.id)
+        }
+        userRoundTicket.amount += amount
+        userRoundTicket.save()
+    }
+    round.save()
```
Comment (wfubuild) [inline src/Bucket.ts:40]: do we need bucket.save() as well?

File: src/Bucket.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        let id = round.id + '0'
+        let lastTicket = SearchTicket.load(id)
+        let winners = new RoundWinner(round.id+i.toString())
+        winners.round = round.id
+        winners.winner = lastTicket.user
+        winners.numberWhitelist = i;
+        winners.save()
+        round.result.push(winners.id)
+    }
+    round.save()
```
Comment (wfubuild) [inline src/Bucket.ts:61]: same here, do we need bucket.save() ?

**Comment by JackyWYX** [comment]
Did we reach an agreement that we may remove `selectedTickets`?

Also, please resolve the comments please. Thanks!

File: README.md (hunk 7 above)
Comment (JackyWYX) [reply README.md:13]: @wfubuild Actually typechain is the auto-generated data from `yarn compile`. Shall always keep typechain in gitignore file.

**Pull request merged**